src/Elliptic/DiscontinuousGalerkin/NumericalFluxes/FirstOrderInternalPenalty.hpp

```diff
@@ -0,0 +1,379 @@
+ * don't depend on the dynamic variables (which may be discontinuous on element
+ * faces). This is the case for the problems we are expecting to solve, because
+ * those will be on fixed background metrics (e.g. a conformal metric for the
+ * XCTS system).
+ *
+ * Also note that the numerical fluxes intentionally don't depend on the
+ * auxiliary field values \f$v\f$. This property allows us to use the numerical
+ * fluxes also for the second-order (or _primal_) DG formulation, where we
+ * remove the need for an auxiliary variable. For the first-order system we
+ * could replace the divergence in \f$F^{*i}_u\f$ with \f$v\f$, which would
+ * result in a generalized _stabilized central flux_. It resembles the
+ * Local-Lax-Friedrichs flux for evolution systems but remains consistent in the
+ * stabilizer term even for principal parts that depend on dynamic variables.
+ *
+ * For a Poisson system \f$-\Delta u(x) = f(x)\f$ with auxiliary variable
+ * \f$v_i=\partial_i u\f$ this numeric flux reduces to the standard internal
+ * penalty flux (see e.g. \cite HesthavenWarburton, section 7.2, or
+ * \cite Arnold2002)
+ *
+ * \f{align}
+ * n_i F^{*i}_u = n_i v_i^* = \frac{1}{2} n_i \left(\partial_i u^\mathrm{int} +
+ * \partial_i u^\mathrm{ext}\right) - \sigma \left(u^\mathrm{int} -
+ * u^\mathrm{ext}\right)
+ * n_i F^{*i}_{v_j} = n_j u^* = \frac{1}{2} n_j \left(u^\mathrm{int} +
+ * u^\mathrm{ext}\right)
+ * \f}
+ *
+ * where a sum over repeated indices is assumed, since the equation is
+ * formulated on a Euclidean geometry.
+ *
+ * This generalization of the internal penalty flux is based on unpublished work
+ * by Nils L. Fischer (nils.fischer@aei.mpg.de).
+ *
+ * The penalty factor \f$\sigma\f$ is responsible for removing zero eigenmodes
+ * and impacts the conditioning of the linear operator to be solved. It can be
+ * chosen as \f$\sigma=C\frac{N_\mathrm{points}^2}{h}\f$ where
+ * \f$N_\mathrm{points}\f$ is the number of collocation points (i.e. the
+ * polynomial degree plus 1), \f$h\f$ is a measure of the element size in
+ * inertial coordinates and \f$C\leq 1\f$ is a free parameter (see e.g.
+ * \cite HesthavenWarburton, section 7.2).
+ */
+template <size_t Dim, typename FluxesComputerTag, typename FieldTagsList,
+          typename AuxiliaryFieldTagsList,
+          typename FluxesComputer = db::item_type<FluxesComputerTag>,
+          typename FluxesArgs = typename FluxesComputer::argument_tags>
+struct FirstOrderInternalPenalty;
+
+template <size_t Dim, typename FluxesComputerTag, typename... FieldTags,
+          typename... AuxiliaryFieldTags, typename FluxesComputer,
+          typename... FluxesArgs>
+struct FirstOrderInternalPenalty<Dim, FluxesComputerTag,
+                                 tmpl::list<FieldTags...>,
+                                 tmpl::list<AuxiliaryFieldTags...>,
+                                 FluxesComputer, tmpl::list<FluxesArgs...>> {
```

Comment:
C++ newbie question: Can you please explain what's going on here, so I can understand? Why do you need to declare the struct taking lists as parameters and then use that on line 124 when you declare the struct with variadic parameters, instead of just declaring the struct using variadic parameters in the first place?

Reply:
This is to expand the lists into a template parameter pack. We generally work with lists since they're easier to handle and pass around. But for the implementation of this class I want to generate function signatures and such from the elements of the lists, and for that I need to expand them into a parameter pack.